04-spiral_classification.ipynb

```diff
@@ -308,6 +308,78 @@
    "source": [
     "# ! ffmpeg -i {m}-3d-levels/%03d.png -vf \"crop=trunc(iw/2)*2:trunc(ih/2)*2\" -pix_fmt yuv420p {m}-3d-levels.mp4"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Low dimensional embeddings"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model = nn.Sequential(\n",
+    "    nn.Linear(n, d),\n",
+    "    nn.ReLU(),\n",
+    "    nn.Linear(d, 2),  # new layer with output in R^2\n",
+    "    nn.Linear(2, K)\n",
+    ")\n",
+    "model.to(device)\n",
+    "\n",
+    "C = torch.nn.CrossEntropyLoss(reduction='none')\n",
+    "optimiser = torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=lambda_l2)\n",
+    "\n",
+    "for t in range(2_000):\n",
+    "    l = model(X)\n",
+    "    F = C(l, y)\n",
+    "    L = F.mean()\n",
+    "    optimiser.zero_grad()\n",
+    "    L.backward()\n",
+    "    optimiser.step()\n",
+    "    overwrite(f'[EPOCH]: {t}, [LOSS]: {L.item():.6f}, [ACCURACY]: {acc(l, y):.3f}')\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Use forward hook to get 2D embeddings (output of nn.Linear(d, 2) in the model) \n",
+    "layer_outputs = {}\n",
+    "\n",
+    "def get_layer_outputs(name):\n",
+    "    def hook(model, input, output):\n",
+    "        layer_outputs[name] = output.detach()\n",
+    "    return hook\n",
+    "\n",
+    "model[-2].register_forward_hook(get_layer_outputs('low_dim_embeddings'))\n",
+    "\n",
+    "# pass data through model to populate layer_outputs\n",
+    "output = model(X)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%matplotlib inline\n",
+    "set_default()\n",
```

Comment:
Do we need these?

Reply:
Previously yes (at least with my local config, as the code was after `%matplotlib notebook`). 
This is now removed in the PR update :)